Ensure consistent drop for panicking drop in hint::select_unpredictable

197g opened the pull request:
There are a few alternatives to the implementation. The principal problem is that the selected value must be owned (in the sense of having a drop flag of sorts) when the unselected value is dropped, such that panic unwind goes through the drop of both. This ownership must then be passed on in return when the drop went smoothly. 


The basic way of achieving this is by extracting the selected value first, at the cost of relying on the optimizer a little more for detecting the copy as constructing the return value despite having a place in the body. Unfortunately, that causes LLVM to discard the !unpredictable annotation (for some reason that is beyond my comprehension of LLVM).

<details>
<summary>Extract from the build log showing an unannotated select being used</summary>

```
2025-08-09T16:51:06.8790764Z            39: define noundef i64 @test_int2(i1 noundef zeroext %p, i64 noundef %a, i64 noundef %b) unnamed_addr #0 personality ptr @rust_eh_personality { 
2025-08-09T16:51:06.8791368Z check:47'0                                  X~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ error: no match found
2025-08-09T16:51:06.8791700Z            40: start: 
2025-08-09T16:51:06.8791858Z check:47'0     ~~~~~~~
2025-08-09T16:51:06.8792043Z            41:  %ret.i = select i1 %p, i64 %a, i64 %b 
2025-08-09T16:51:06.8792293Z check:47'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
2025-08-09T16:51:06.8792686Z check:47'1               ?                             possible intended match
2025-08-09T16:51:06.8792946Z            42:  ret i64 %ret.i 
2025-08-09T16:51:06.8793127Z check:47'0     ~~~~~~~~~~~~~~~~
```

</details>

So instead, this PR includes a guard to drop the selected `MaybeUnit<T>` which is active only for the section where the unselected value is dropped. That leaves the code for selecting the result intact leading to the expected ir. That complicates the 'unselection' process a little bit since we require _both_ values as a result of that intrinsic call. Since the arguments alias, this portion as well as the drop guard uses raw pointers.

Closes: rust-lang/rust#145148
Prior: rust-lang/rust#139977

197g committed:
Ensure consistent drop for hint::select_unpredictable

There are a few alternatives to the implementation. The principal
problem is that the selected value must be owned (in the sense of having
any drop flag of sorts) when the unselected value is dropped, such that
panic unwind goes through the drop of both. This ownership must then be
passed on in return when the drop went smoothly. The basic way of
achieving this is by extracting the selected value first, at the cost of
relying on the optimizer a little more for detecting the copy as
constructing the return value despite having a place in the body.

rust-log-analyzer commented:
The job **`aarch64-gnu-llvm-19-1`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/16851285156/job/47737917992) [(plain enhanced)](https://triage.rust-lang.org/gha-logs/rust-lang/rust/47737917992) [(plain)](https://github.com/rust-lang/rust/commit/21bdfa14cbd99aba173751a2c723fb1e65de401b/checks/47737917992/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
failures:

---- [codegen] tests/codegen-llvm/intrinsics/select_unpredictable.rs stdout ----

error: verification with 'FileCheck' failed
status: exit status: 1
command: "/usr/lib/llvm-19/bin/FileCheck" "--input-file" "/checkout/obj/build/aarch64-unknown-linux-gnu/test/codegen-llvm/intrinsics/select_unpredictable/select_unpredictable.ll" "/checkout/tests/codegen-llvm/intrinsics/select_unpredictable.rs" "--check-prefix=CHECK" "--allow-unused-prefixes" "--dump-input-context" "100"
stdout: none
--- stderr -------------------------------
/checkout/tests/codegen-llvm/intrinsics/select_unpredictable.rs:47:12: error: CHECK: expected string not found in input
 // CHECK: select i1 %p, i64 %a, i64 %b, !unpredictable
           ^
/checkout/obj/build/aarch64-unknown-linux-gnu/test/codegen-llvm/intrinsics/select_unpredictable/select_unpredictable.ll:39:30: note: scanning from here
define noundef i64 @test_int2(i1 noundef zeroext %p, i64 noundef %a, i64 noundef %b) unnamed_addr #0 personality ptr @rust_eh_personality {
                             ^
/checkout/obj/build/aarch64-unknown-linux-gnu/test/codegen-llvm/intrinsics/select_unpredictable/select_unpredictable.ll:41:11: note: possible intended match here
 %ret.i = select i1 %p, i64 %a, i64 %b
          ^
/checkout/tests/codegen-llvm/intrinsics/select_unpredictable.rs:54:12: error: CHECK: expected string not found in input
 // CHECK: select i1 %p, {{.*}}, !unpredictable
           ^
/checkout/obj/build/aarch64-unknown-linux-gnu/test/codegen-llvm/intrinsics/select_unpredictable/select_unpredictable.ll:46:32: note: scanning from here
define { i64, i64 } @test_pair2(i1 noundef zeroext %p, i64 noundef %a.0, i64 noundef %a.1, i64 noundef %b.0, i64 noundef %b.1) unnamed_addr #0 personality ptr @rust_eh_personality {
                               ^
/checkout/obj/build/aarch64-unknown-linux-gnu/test/codegen-llvm/intrinsics/select_unpredictable/select_unpredictable.ll:48:29: note: possible intended match here
 %ret.0.i.sroa.speculated = select i1 %p, i64 %a.0, i64 %b.0
                            ^

Input file: /checkout/obj/build/aarch64-unknown-linux-gnu/test/codegen-llvm/intrinsics/select_unpredictable/select_unpredictable.ll
Check file: /checkout/tests/codegen-llvm/intrinsics/select_unpredictable.rs

-dump-input=help explains the following input dump.

Input was:
<<<<<<
            1: ; ModuleID = 'select_unpredictable.e9f1accfe315d8ff-cgu.0' 
            2: source_filename = "select_unpredictable.e9f1accfe315d8ff-cgu.0" 
            3: target datalayout = "e-m:e-i8:8:32-i16:16:32-i64:64-i128:128-n32:64-S128-Fn32" 
            4: target triple = "aarch64-unknown-linux-gnu" 
            5:  
            6: ; Function Attrs: mustprogress nofree norecurse nosync nounwind willreturn memory(none) uwtable 
            7: define noundef i64 @test_int(i1 noundef zeroext %p, i64 noundef %a, i64 noundef %b) unnamed_addr #0 { 
            8: start: 
            9:  %0 = select i1 %p, i64 %a, i64 %b, !unpredictable !2 
           10:  ret i64 %0 
           11: } 
           12:  
           13: ; Function Attrs: mustprogress nofree norecurse nosync nounwind willreturn memory(none) uwtable 
           14: define { i64, i64 } @test_pair(i1 noundef zeroext %p, i64 noundef %a.0, i64 noundef %a.1, i64 noundef %b.0, i64 noundef %b.1) unnamed_addr #0 { 
           15: start: 
           16:  %0 = insertvalue { i64, i64 } poison, i64 %a.0, 0 
           17:  %1 = insertvalue { i64, i64 } %0, i64 %a.1, 1 
           18:  %2 = insertvalue { i64, i64 } poison, i64 %b.0, 0 
           19:  %3 = insertvalue { i64, i64 } %2, i64 %b.1, 1 
           20:  %4 = select i1 %p, { i64, i64 } %1, { i64, i64 } %3, !unpredictable !2 
           21:  ret { i64, i64 } %4 
           22: } 
           23:  
           24: ; Function Attrs: mustprogress nofree norecurse nosync nounwind willreturn memory(read, argmem: readwrite, inaccessiblemem: none) uwtable 
           25: define void @test_struct(ptr dead_on_unwind noalias nocapture noundef writable writeonly sret([800 x i8]) align 8 dereferenceable(800) %_0, i1 noundef zeroext %p, ptr noalias nocapture noundef readonly align 8 dereferenceable(800) %a, ptr noalias nocapture noundef readonly align 8 dereferenceable(800) %b) unnamed_addr #1 { 
           26: start: 
           27:  %0 = select i1 %p, ptr %a, ptr %b, !unpredictable !2 
           28:  tail call void @llvm.memcpy.p0.p0.i64(ptr noundef nonnull align 8 dereferenceable(800) %_0, ptr noundef nonnull align 8 dereferenceable(800) %0, i64 800, i1 false) 
           29:  ret void 
           30: } 
           31:  
           32: ; Function Attrs: mustprogress nofree norecurse nosync nounwind willreturn memory(none) uwtable 
           33: define void @test_zst(i1 noundef zeroext %p) unnamed_addr #0 { 
           34: start: 
           35:  ret void 
           36: } 
           37:  
           38: ; Function Attrs: mustprogress nofree norecurse nosync nounwind willreturn memory(none) uwtable 
           39: define noundef i64 @test_int2(i1 noundef zeroext %p, i64 noundef %a, i64 noundef %b) unnamed_addr #0 personality ptr @rust_eh_personality { 
check:47'0                                  X~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ error: no match found
           40: start: 
check:47'0     ~~~~~~~
           41:  %ret.i = select i1 %p, i64 %a, i64 %b 
check:47'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
check:47'1               ?                             possible intended match
           42:  ret i64 %ret.i 
check:47'0     ~~~~~~~~~~~~~~~~
           43: } 
check:47'0     ~~
           44:  
check:47'0     ~
           45: ; Function Attrs: mustprogress nofree norecurse nosync nounwind willreturn memory(none) uwtable 
check:47'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           46: define { i64, i64 } @test_pair2(i1 noundef zeroext %p, i64 noundef %a.0, i64 noundef %a.1, i64 noundef %b.0, i64 noundef %b.1) unnamed_addr #0 personality ptr @rust_eh_personality { 
check:47'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
check:54'0                                    X~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ error: no match found
           47: start: 
check:54'0     ~~~~~~~
           48:  %ret.0.i.sroa.speculated = select i1 %p, i64 %a.0, i64 %b.0 
check:54'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
check:54'1                                 ?                                 possible intended match
           49:  %ret.1.i.sroa.speculated = select i1 %p, i64 %a.1, i64 %b.1 
check:54'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           50:  %0 = insertvalue { i64, i64 } poison, i64 %ret.0.i.sroa.speculated, 0 
check:54'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           51:  %1 = insertvalue { i64, i64 } %0, i64 %ret.1.i.sroa.speculated, 1 
check:54'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           52:  ret { i64, i64 } %1 
check:54'0     ~~~~~~~~~~~~~~~~~~~~~
           53: } 
check:54'0     ~~
           54:  
check:54'0     ~
           55: ; Function Attrs: mustprogress nofree norecurse nosync nounwind willreturn memory(read, argmem: readwrite, inaccessiblemem: none) uwtable 
check:54'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           56: define void @test_struct2(ptr dead_on_unwind noalias nocapture noundef writable writeonly sret([800 x i8]) align 8 dereferenceable(800) %_0, i1 noundef zeroext %p, ptr noalias nocapture noundef readonly align 8 dereferenceable(800) %a, ptr noalias nocapture noundef readonly align 8 dereferenceable(800) %b) unnamed_addr #1 personality ptr @rust_eh_personality { 
check:54'0     ~~~~~~~~~~~~~~~~~~~~~~~~~
           57: start: 
           58:  %0 = select i1 %p, ptr %a, ptr %b, !unpredictable !2 
           59:  tail call void @llvm.memcpy.p0.p0.i64(ptr noundef nonnull align 8 dereferenceable(800) %_0, ptr noundef nonnull align 8 dereferenceable(800) %0, i64 800, i1 false), !noalias !3 
           60:  ret void 
           61: } 
           62:  
           63: ; Function Attrs: mustprogress nofree norecurse nosync nounwind willreturn memory(none) uwtable 
           64: define void @test_zst2(i1 noundef zeroext %p) unnamed_addr #0 personality ptr @rust_eh_personality { 
           65: start: 
           66:  ret void 
           67: } 
           68:  
           69: ; Function Attrs: nounwind uwtable 
           70: declare noundef range(i32 0, 10) i32 @rust_eh_personality(i32 noundef, i32 noundef, i64 noundef, ptr noundef, ptr noundef) unnamed_addr #2 
           71:  
           72: ; Function Attrs: mustprogress nocallback nofree nounwind willreturn memory(argmem: readwrite) 
           73: declare void @llvm.memcpy.p0.p0.i64(ptr noalias nocapture writeonly, ptr noalias nocapture readonly, i64, i1 immarg) #3 
           74:  
           75: attributes #0 = { mustprogress nofree norecurse nosync nounwind willreturn memory(none) uwtable "frame-pointer"="non-leaf" "probe-stack"="inline-asm" "target-cpu"="generic" "target-features"="+v8a,+outline-atomics" } 
           76: attributes #1 = { mustprogress nofree norecurse nosync nounwind willreturn memory(read, argmem: readwrite, inaccessiblemem: none) uwtable "frame-pointer"="non-leaf" "probe-stack"="inline-asm" "target-cpu"="generic" "target-features"="+v8a,+outline-atomics" } 
           77: attributes #2 = { nounwind uwtable "frame-pointer"="non-leaf" "probe-stack"="inline-asm" "target-cpu"="generic" "target-features"="+v8a,+outline-atomics" } 
           78: attributes #3 = { mustprogress nocallback nofree nounwind willreturn memory(argmem: readwrite) } 
           79:  
           80: !llvm.module.flags = !{!0} 
           81: !llvm.ident = !{!1} 
           82:  
           83: !0 = !{i32 8, !"PIC Level", i32 2} 
           84: !1 = !{!"rustc version 1.91.0-nightly (00fc429ff 2025-08-09)"} 
           85: !2 = !{} 
           86: !3 = !{!4, !6} 
           87: !4 = distinct !{!4, !5, !"_ZN4core4hint20select_unpredictable17h598d305da1153d4fE: %true_val"} 
           88: !5 = distinct !{!5, !"_ZN4core4hint20select_unpredictable17h598d305da1153d4fE"} 
           89: !6 = distinct !{!6, !5, !"_ZN4core4hint20select_unpredictable17h598d305da1153d4fE: %false_val"} 
>>>>>>
------------------------------------------



```

</details>

197g committed:
Adjust for codegen of unpredictable return value

It seems important for LLVM that we select on the values by-value
instead of reading and have no intermediate store. So make sure the
guards selects both potential drops but defers the return value to the
second selection. Since the two options alias we use raw mutable
pointers instead of mutable references as before.

joboet commented on library/core/src/hint.rs:
```diff
@@ -786,12 +786,42 @@ pub fn select_unpredictable<T>(condition: bool, true_val: T, false_val: T) -> T
     let mut false_val = MaybeUninit::new(false_val);
+
+    struct DropOnPanic<T> {
+        inner: *mut MaybeUninit<T>,
```
References should be fine here, right?

```suggestion
        inner: &'a mut MaybeUninit<T>,
```

197g replied:
We have the same two pointers in both arguments to the first `select_unpredictable`, each one duplicate. So no.

joboet replied:
Oh right, fair enough!

197g replied:
Do you think it's simpler code to dereference the pointer as it gets constructed into a `Guard`? I thought since we're calling `MaybeUninit`'s proxy to `drop_in_place` it would just add one additional location of unsafe precondition,

197g replied:
Another option would be to avoid the `MaybeUninit` entirely in the type and use just raw pointers with direct `drop_in_place` calls. The `assume_init_drop` is noise with an additional precondition that must be fulfilled anyways. And if the comment of moving to `ManuallyDrop` is eventually followed-up on those attributes would remain untouched.

joboet replied:
> Another option would be to avoid the `MaybeUninit` entirely in the type and use just raw pointers with direct `drop_in_place` calls. The `assume_init_drop` is noise with an additional precondition that must be fulfilled anyways. And if the comment of moving to `ManuallyDrop` is eventually followed-up on those attributes would remain untouched.

Personally, I think this might be the cleanest option.

joboet commented on library/core/src/hint.rs:
```diff
@@ -786,12 +786,42 @@ pub fn select_unpredictable<T>(condition: bool, true_val: T, false_val: T) -> T
-            .assume_init_drop();
+        // Extract the selected value first, ensure it is dropped as well if dropping the unselected
+        // value panics.
+        let (guard, drop) = crate::intrinsics::select_unpredictable(
+            condition,
+            (true_ptr, false_ptr),
+            (false_ptr, true_ptr),
+        );
```
I'd split this up into two `select_unpredictable`s – that probably yields simpler LLVM IR than the `insertvalue` and `extractvalue` calls of this version.

joboet commented on library/coretests/tests/hint.rs:
```diff
@@ -21,3 +21,40 @@ fn select_unpredictable_drop() {
+    //
+    // If `selected` is not dropped, `armed` panics as 1 != 0xdead
+    let _unreachable =
+        core::hint::select_unpredictable(core::hint::black_box(true), selected, unselected);
```
`black_box` is unnecessary here, this isn't testing the codegen nor the performance.
```suggestion
        core::hint::select_unpredictable(true, selected, unselected);
```

197g committed:
Clarify panic-drop test for select_unpredictable

197g committed:
Simplify select_unpredictable guard selection

Instead of a tuple, select the dropped value and its guard with two
separate calls to the intrinsic which makes both calls have a
pointer-valued argument that should be simpler in codegen. Use the same
condition on all (not an inverted condition) to clarify the intent of
parallel selection. This should also be a simpler value-dependency chain
if the guard is deduced unused (i.e. drop_in_place a noop for the type).

197g committed:
Switch select_unpredictable guard to raw pointer

joboet commented:
@bors r+

bors commented:
:pushpin: Commit ee9803a244d65a830a999f8066da2a70e1972d5e has been approved by `joboet`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=joboet ee9803a244d65a830a999f8066da2a70e1972d5e -->
<!-- homu: {"type":"Approved","sha":"ee9803a244d65a830a999f8066da2a70e1972d5e","approver":"joboet","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
